### PR TITLE
Code quality test

### PR DESCRIPTION
**Pull request by astanziola**
Adds a separate CI test for MATLAB linting errors using `checkcode`. 
It also separates it from the remaining of the test suite, although I'm not sure that's done in the best possible way.

### CODE REVIEW

**Comment by astanziola** [comment]
@dstansby This should be ready, if you're happy to review it.

I've also added another workflow for code linting, separated from the other tests, which needs your approval before being run the first time.

File: linting/code-quality/code_quality_test.m [hunk 1]
```diff
@@ -0,0 +1,78 @@
+% Linter warning prototype
+
+classdef code_quality_test < matlab.unittest.TestCase
+
+  methods
+
+    function serious_problems = filterProblems(testCase, problems, mfile)
+      % Some problems should only be warnings, not errors.
+      % This function filters them out.
+      % You can add more problems to the list if you want.
```
Comment (dstansby) [inline linting/code-quality/code_quality_test.m]: Could you improve this comment to note whether warnings fail CI or not?

File: run_code_quality.m [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import matlab.unittest.Verbosity
+
+
+oldpath = path;
+path('.', oldpath);
+
+code_quality_path = fullfile(pwd, 'linting', 'code-quality');
+suite = testsuite(code_quality_path, 'IncludeSubfolders', true);
+
+[~,~] = mkdir('artifacts');
```
Comment (dstansby) [inline run_code_quality.m]: If we're not saving any code coverage output (which we shouldn't with linting!) I don't think this directory needs creating?

**Review by dstansby — COMMENTED** [review]
Looks 👍 overall. Apart from my other inline comments I guess we want to get rid of `sloppy_fun.m`  before merging?

File: linting/code-quality/code_quality_test.m (hunk 1 above)
Comment (astanziola) [reply linting/code-quality/code_quality_test.m]: Good observation! Does it look good now?

File: run_code_quality.m (hunk 2 above)
Comment (astanziola) [reply run_code_quality.m]: Yes! Just removed it

**Comment by astanziola** [comment]
@dstansby Can I ask for an advice?
Not sure what's the best way to handle the conflicts here: shall I just rebase this into the main, or is there a better way to fix diverging branches than rewriting the history?

**Comment by dstansby** [comment]
Rebasing (and resolve the conflicts) is fine, and what I usually do (mainly out of habit, but also to keep a linear git history. The other option is to take the current `main` branch and merge it into your branch (and resolve the conflicts). That won't re-write any of your previous commits, but instead create a new commit on top of them. I'd say do whichever you're more comfortable doing, as the result is the same 😃

**Comment by dstansby** [comment]
Just thinking about this again, I think the best thing to do would be revert changes to the `.m` files which are conflicting, then you should be able to do a clean merge of `main` into this branch (and then probably one more commit to fix any linting issues I've introduced since you opened this PR 😉 )

**Comment by dstansby** [comment]
Going to close and open to see if the unit tests will run properley

**Comment by dstansby** [comment]
Not sure why the unit tests aren't running. I'll merge this without, and if there are issue on merge I'll fix them in another PR. Thanks a lot @astanziola for this 👍

**Pull request merged**